pallets/pallet-asset-swap/src/lib.rs

```diff
@@ -0,0 +1,548 @@
+// KILT Blockchain – https://botlabs.org
+// Copyright (C) 2019-2024 BOTLabs GmbH
+
+// The KILT Blockchain is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The KILT Blockchain is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+// If you feel like getting in touch with us, you can do so at info@botlabs.org
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+pub mod xcm;
+
+mod swap;
+pub use swap::{SwapPairInfo, SwapPairStatus};
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+
+use ::xcm::{VersionedAssetId, VersionedMultiAsset, VersionedMultiLocation};
+use parity_scale_codec::{Decode, Encode};
+use sp_runtime::traits::TrailingZeroInput;
+use sp_std::boxed::Box;
+
+pub use crate::pallet::*;
+
+const LOG_TARGET: &str = "runtime::pallet-asset-swap";
+
+#[frame_support::pallet(dev_mode)]
+pub mod pallet {
+	use crate::{
+		swap::{SwapPairInfo, SwapPairStatus},
+		LOG_TARGET,
+	};
+
+	use frame_support::{
+		pallet_prelude::*,
+		traits::{
+			fungible::{Inspect as InspectFungible, Mutate as MutateFungible},
+			tokens::{Fortitude, Preservation, Provenance},
+			EnsureOrigin,
+		},
+	};
+	use frame_system::{ensure_root, pallet_prelude::*};
+	use sp_runtime::traits::TryConvert;
+	use sp_std::{boxed::Box, vec};
+	use xcm::{
+		v3::{
+			validate_send, AssetId,
+			Instruction::{BuyExecution, DepositAsset, RefundSurplus, SetAppendix, TransferAsset, WithdrawAsset},
+			Junction, MultiAsset, MultiAssetFilter, MultiLocation, SendXcm, WeightLimit, WildMultiAsset, Xcm,
+		},
+		VersionedAssetId, VersionedMultiAsset, VersionedMultiLocation,
+	};
+	use xcm_executor::traits::TransactAsset;
+
+	pub type LocalCurrencyBalanceOf<T> =
+		<<T as Config>::LocalCurrency as InspectFungible<<T as frame_system::Config>::AccountId>>::Balance;
+	pub type SwapPairInfoOf<T> = SwapPairInfo<<T as frame_system::Config>::AccountId>;
+
+	const STORAGE_VERSION: StorageVersion = StorageVersion::new(0);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		const POOL_ADDRESS_GENERATION_ENTROPY: [u8; 8];
+
+		type AccountIdConverter: TryConvert<Self::AccountId, Junction>;
```

Comment:
Could you please explain, why this is needed? This feels to me somehow overengineered

Reply:
It does feel, but then if you think about what the definition of `AccountId` is for `frame_system`, you realise you cannot assume it will always be `AccountId32`. For instance, a parachain using `AccountId20` would need to use a different converter. If we bake in the assumption that an account ID is always an `AccountId32`, then the pallet is not usable anywhere else.